### Loading the dependencies

In [1]:
import pandas as pd
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

### Loading the hate crime dataset

In [2]:
df = pd.read_csv("hate_crime.csv", low_memory=False)
df.shape

(219577, 28)

In [3]:
df.head().T

0  \
INCIDENT_ID                                             3015   
DATA_YEAR                                               1991   
ORI                                                AR0040200   
PUB_AGENCY_NAME                                       Rogers   
PUB_AGENCY_UNIT                                          NaN   
AGENCY_TYPE_NAME                                        City   
STATE_ABBR                                                AR   
STATE_NAME                                          Arkansas   
DIVISION_NAME                             West South Central   
REGION_NAME                                            South   
POPULATION_GROUP_CODE                                      5   
POPULATION_GROUP_DESC         Cities from 10,000 thru 24,999   
INCIDENT_DATE                                      31-AUG-91   
ADULT_VICTIM_COUNT                                       NaN   
JUVENILE_VICTIM_COUNT                                    NaN   
TOTAL_OFFENDER_COUNT                                       1   
ADULT_OFFENDER_COUNT                                     NaN   
JUVENILE_OFFENDER_COUNT                                  NaN   
OFFENDER_RACE                                          White   
OFFENDER_ETHNICITY                                       NaN   
VICTIM_COUNT                                               1   
OFFENSE_NAME                                    Intimidation   
TOTAL_INDIVIDUAL_VICTIMS                                   1   
LOCATION_NAME             Highway/Road/Alley/Street/Sidewalk   
BIAS_DESC                     Anti-Black or African American   
VICTIM_TYPES                                      Individual   
MULTIPLE_OFFENSE                                           S   
MULTIPLE_BIAS                                              S   

                                                           1  \
INCIDENT_ID                                             3016   
DATA_YEAR                                               1991   
ORI                                                AR0290100   
PUB_AGENCY_NAME                                         Hope   
PUB_AGENCY_UNIT                                          NaN   
AGENCY_TYPE_NAME                                        City   
STATE_ABBR                                                AR   
STATE_NAME                                          Arkansas   
DIVISION_NAME                             West South Central   
REGION_NAME                                            South   
POPULATION_GROUP_CODE                                      6   
POPULATION_GROUP_DESC           Cities from 2,500 thru 9,999   
INCIDENT_DATE                                      19-SEP-91   
ADULT_VICTIM_COUNT                                       NaN   
JUVENILE_VICTIM_COUNT                                    NaN   
TOTAL_OFFENDER_COUNT                                       1   
ADULT_OFFENDER_COUNT                                     NaN   
JUVENILE_OFFENDER_COUNT                                  NaN   
OFFENDER_RACE                      Black or African American   
OFFENDER_ETHNICITY                                       NaN   
VICTIM_COUNT                                               1   
OFFENSE_NAME                                  Simple Assault   
TOTAL_INDIVIDUAL_VICTIMS                                   1   
LOCATION_NAME             Highway/Road/Alley/Street/Sidewalk   
BIAS_DESC                                         Anti-White   
VICTIM_TYPES                                      Individual   
MULTIPLE_OFFENSE                                           S   
MULTIPLE_BIAS                                              S   

                                                       2  \
INCIDENT_ID                                           43   
DATA_YEAR                                           1991   
ORI                                            AR0350100   
PUB_AGENCY_NAME                               Pine Bluff   
PUB_AGENCY_UNIT                              

In [4]:
# Converting column names to lower case for convenience purpose
df= df.rename(columns=str.lower)

### Changing the data type of date values

In [5]:
df['incident_date'] = pd.to_datetime(df['incident_date'], format = '%d-%b-%y')

In [6]:
# Check the percentages of missing value
df.isnull().sum() / df.shape[0] * 100

incident_id                  0.000000
data_year                    0.000000
ori                          0.000000
pub_agency_name              0.000000
pub_agency_unit             97.071187
agency_type_name             0.000000
state_abbr                   0.000000
state_name                   0.000000
division_name                0.000000
region_name                  0.000000
population_group_code        0.000000
population_group_desc        0.000000
incident_date                0.000000
adult_victim_count          76.586346
juvenile_victim_count       77.438894
total_offender_count         0.000000
adult_offender_count        78.536003
juvenile_offender_count     78.539191
offender_race                0.009108
offender_ethnicity          84.118100
victim_count                 0.000000
offense_name                 0.000000
total_individual_victims     1.584410
location_name                0.000000
bias_desc                    0.000000
victim_types                 0.000000
multiple_off

Columns with a very high percentage of missing values seem to be unimportant for tha analysis. So, they are going to be dropped from the dataframe.

The column 'total_individual_victims' has 1.58% missing values but it also seems unnecessary so dropping that column as well.

In [7]:
df.drop(['pub_agency_unit','adult_victim_count','juvenile_victim_count','adult_offender_count',\
         'juvenile_offender_count','offender_ethnicity'], axis=1, inplace=True)
df.drop(['total_individual_victims'], axis=1, inplace=True)

In [12]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 219577 entries, 0 to 219576
Data columns (total 21 columns):
 #   Column                 Non-Null Count   Dtype         
---  ------                 --------------   -----         
 0   incident_id            219577 non-null  int64         
 1   data_year              219577 non-null  int64         
 2   ori                    219577 non-null  object        
 3   pub_agency_name        219577 non-null  object        
 4   agency_type_name       219577 non-null  object        
 5   state_abbr             219577 non-null  object        
 6   state_name             219577 non-null  object        
 7   division_name          219577 non-null  object        
 8   region_name            219577 non-null  object        
 9   population_group_code  219577 non-null  object        
 10  population_group_desc  219577 non-null  object        
 11  incident_date          219577 non-null  datetime64[ns]
 12  total_offender_count   219577 non-null  int6

## Answers of the research questions

In [21]:
df.groupby(['state_name','data_year'==2020]).sum()

KeyError: False